### ISSUE TITLE
[feature] Generator for secrets

### ISSUE THREAD

**Issue by jastBytes** [external-secrets/external-secrets#1179]
**Describe the solution you'd like**

There are use-cases where a secret could be generated by the operator instead of using pre-existing ones from the secret store. For example a MongoDB could be spun up using a password generated by the operator. This password is then used by an application in the cluster to access the database. No human ever needs to read that secret or have to specify it beforehand. A similar feature is available in a project called [`VaultOperator`](https://github.com/finleap-connect/vaultoperator) which tries to solve similar problems like this operator just for Hashicorp Vault as a secret store.

**What is the added value?**

Secrets can be generated automatically without manual interaction with the secret store.

**Give us examples of the outcome**

Example of generators:
- `string` generates a random string where length and character set can be specified.
- `bytes` generates random bytes with a given length

**Comment by gusfcarvalho** [comment]
Hi @jastBytes ! In the past we already discussed about it (https://github.com/external-secrets/external-secrets/issues/249#issuecomment-953152407), and we don't think creating secrets is something we wanted to go to (then again, we are going for sinking secrets).

**Comment by jastBytes** [comment]
I respect the decision not to have such a feature to keep it clean and straight pulling secrets from secret stores. Thanks for fast response.

**Comment by gusfcarvalho** [comment]
Maybe the other maintainers have new thoughts on it. @moolen @knelasevero

**Comment by gusfcarvalho** [comment]
Myself, I still think this is too much to handle  (e.g. rotations, etc).

**Comment by moolen** [comment]
hey Jan @jastBytes :wave: i hope you enjoyed the rest of kubecon :)

Hmm, i think from a product perspective it fits quite nicely with external-secrets. The assumption in the above mentioned comment was, that users create infrastructure with crossplane or with an operator and then push the generated secrets to some provider. But people should have a tool to generate secrets in Kubernetes and provide them to a helm chart or so.

I'm open to add that, but we need to discover and document the use-cases before we go into design and implementation.
We can definitely take some parts from the `generator` of your https://github.com/finleap-connect/vaultoperator.

From the top of my head we can support three types of operation, each can have its own CRD:
1. fetching secrets
2. pushing secrets
3. generating secrets

While 2+3 are somewhat interleaved, we can consider putting the `generator` functionality on the push CRD to make it more concise.

**Comment by moolen** [comment]
We discussed this issue today in the community meeting:
tl;dr yes, please! But we lack capacity for this RN :disappointed: 

So, tbh there's so much going on with the project right now that we don't have the capacity to work on this right away. We want to focus on (1) stabilizing the current set of features and (2) deliver the ability to push secrets.

For such a quite big changes to the project we employ a [proposal process](https://external-secrets.io/v0.5.3/contributing-process/#proposal-process) to gather feedback around that feature before we implement it. Just to ensure that we understand/explore/document the different use-cases around a particular feature. Feel free to take the lead on that @jastBytes (or whoever reads this :wink:) we will help where we can! If you wanna work on it let us know. 

There is definitely a need for this in the community #249 #195, i'd love to get that in!

**Comment by frimik** [comment]
Glad to see it's on your radar. Came here after looking through the Provider list and not finding a "generator" provider. Would prefer it was provided from external-secrets rather than requiring me to install another controller, even though this one does work fine: https://github.com/mittwald/kubernetes-secret-generator

**Comment by christianhuening** [comment]
I am super happy this is accepted now :)
